### PR TITLE
hovertemplate support for sankey

### LINKED ISSUE #3269: implement `hovertemplate` for Sankey traces
Custom HTML code for both links and nodes.

### PR DESCRIPTION
**Pull request by antoinerg**
Closes #3269

### CODE REVIEW

**Comment by antoinerg** [comment]
I am not sure the description of `hovertemplate` attributes is satisfactory. I'd like to discuss the changes made in commit d952713.

Finally, because the inclusion `hovertemplate` in `Fx.loneHover` calls have been tested in other traces, is it necessary to test it again in `sankey_test.js`? Is it OK to only check for the existence of the expected variables in eventData?

File: src/traces/sankey/attributes.js [hunk 1]
```diff
@@ -146,6 +147,10 @@ var attrs = module.exports = overrideAll({
             ].join(' ')
         },
         hoverlabel: fxAttrs.hoverlabel, // needs editType override,
+        hovertemplate: hovertemplateAttrs({}, {
```
Comment (etpinard) [inline src/traces/sankey/attributes.js]: Hmm. Sankey event data is a bit of mess:

image

image

There's probably way more stuff it in than we need. Moreover, some of it isn't consistent with event data for other traces e.g. the `trace` key should be `fullData`. I'm not how much we can change it w/o making breaking changes.

Oh well, these descriptions look ok to me.

**Comment by etpinard** [comment]
> Finally, because the inclusion `hovertemplate` in `Fx.loneHover` calls have been tested in other traces, is it necessary to test it again in `sankey_test.js`?

I would add one "true" `hovertemplate` test case where we check that a user-defined `hovertemplate` gives the correct hover label content. You're right though, no need to re-test all the `loneHover` logic again :ok_hand:

File: src/traces/sankey/attributes.js (hunk 1 above)
Comment (antoinerg) [reply src/traces/sankey/attributes.js]: I added `fullData` to the eventData in commit 3204274

**Comment by antoinerg** [comment]
> I would add one "true" `hovertemplate` test case where we check that a user-defined `hovertemplate` gives the correct hover label content.

Done in commit 3204274!

File: src/traces/sankey/attributes.js (hunk 1 above)
Comment (etpinard) [reply src/traces/sankey/attributes.js]: v2 changes tracked in https://github.com/plotly/plotly.js/issues/168#issuecomment-315874345

**Comment by etpinard** [comment]
@antoinerg oops I forgot about this PR.

Looks good :dancer:

**Pull request merged**